robot_work.go

```diff
@@ -0,0 +1,200 @@
+package gobot
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"sync"
+
+	"github.com/gobuffalo/uuid"
+)
+
+// RobotWorkRegistry contains all the work units registered on a Robot
+type RobotWorkRegistry struct {
+	sync.RWMutex
+
+	r map[string]*RobotWork
+}
+
+const (
+	EveryWorkKind = "every"
+	AfterWorkKind = "after"
+)
+
+// RobotWork and the RobotWork registry represent units of executing computation
+// managed at the Robot level. Unlike the utility functions gobot.After and gobot.Every,
+// RobotWork units require a context.Context, and can be cancelled externally by calling code.
+//
+// Usage:
+//
+//	someWork := myRobot.Every(context.Background(), time.Second * 2, func(){
+//		fmt.Println("Here I am doing work")
+// 	})
+//
+//	someWork.CallCancelFunc() // Cancel next tick and remove from work registry
+//
+// goroutines for Every and After are run on their own WaitGroups for synchronization:
+//
+//	someWork2 := myRobot.Every(context.Background(), time.Second * 2, func(){
+//		fmt.Println("Here I am doing more work")
+// 	})
+//
+//	somework2.CallCancelFunc()
+//
+//	// wait for both Every calls to finish
+//	robot.WorkEveryWaitGroup().Wait()
+type RobotWork struct {
+	id         uuid.UUID
+	kind       string
+	tickCount  int
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	function   func()
+	ticker     *time.Ticker
+	duration   time.Duration
+}
+
+// ID returns the UUID of the RobotWork
+func (rw *RobotWork) ID() uuid.UUID {
+	return rw.id
+}
+
+// CancelFunc returns the context.CancelFunc used to cancel the work
+func (rw *RobotWork) CancelFunc() context.CancelFunc {
+	return rw.cancelFunc
+}
+
+// CallCancelFunc calls the context.CancelFunc used to cancel the work
+func (rw *RobotWork) CallCancelFunc() {
+	rw.cancelFunc()
+}
+
+// Ticker returns the time.Ticker used in an Every so that calling code can sync on the same channel
+func (rw *RobotWork) Ticker() *time.Ticker {
+	if rw.kind == AfterWorkKind {
+		return nil
+	}
+	return rw.ticker
+}
+
+// TickCount returns the number of times the function successfully ran
+func (rw *RobotWork) TickCount() int {
+	return rw.tickCount
+}
+
+// Duration returns the timeout until an After fires or the period of an Every
+func (rw *RobotWork) Duration() time.Duration {
+	return rw.duration
+}
+
+func (rw *RobotWork) String() string {
+	format := `ID: %s
+Kind: %s
+TickCount: %d
+
+`
+	return fmt.Sprintf(format, rw.id, rw.kind, rw.tickCount)
+}
+
+// WorkRegistry returns the Robot's WorkRegistry
+func (r *Robot) WorkRegistry() *RobotWorkRegistry {
+	return r.workRegistry
+}
+
+// Every calls the given function for every tick of the provided duration.
+func (r *Robot) Every(ctx context.Context, d time.Duration, f func()) *RobotWork {
```

Comment:
Just wondering if this method should be:

```
func (r *Robot) Every(d time.Duration, ctx context.Context, f func()) *RobotWork
```

The idea being that the signature would be a little closer to the `gobot.Every()` function.

Reply:
I'm not dead-set against that, but in most Go code that takes a context, the context is usually passed as the first arg by convention. This still leaves the `time.Duration` and the `func()` next to each other -- potentially it's *more* confusing to have those two split by the new var in the signature, no?

Reply:
Shows how much I know about context. OK, then, let's just carry on with it.

Reply:
😺 I tried to basically follow how it works in the stdlib. Here's an example: https://tip.golang.org/pkg/net/#Dialer.DialContext